raft/src/main/java/org/apache/kafka/raft/LeaderState.java

```diff
@@ -51,6 +53,7 @@ protected LeaderState(
             boolean hasEndorsedLeader = voterId == localId;
```

Comment:
I think we have two options here. This class can contain either:
```java
private final Map<Integer, VoterState> voterReplicaStates = new HashMap<>(); 
```
where we change this expression to `boolean hasEndorsedLeader = grantingVoters.contains(voterId);`
and remove the field `private final Set<Integer> grantingVoters = new HashSet<>();`

or:

```java
private final Set<Integer> voters = new HashSet<>(); 
private final Set<Integer> grantingVoters = new HashSet<>(); 
```
and change a few of the methods here use these two sets instead of `voterReplicaStates`.

I like the first option but it is up to you.

Reply:
@jsancio , even I would prefer option 1 but slightly confused on it. You suggest to check for endorsement using `boolean hasEndorsedLeader = grantingVoters.contains(voterId);` and then asking to remove the field `grantingVoters `. Am I missing something ?

Reply:
Yes, that's correct. You will have to change the implementation of `grantingVoters()` to something like:

```java
public Set<Integer> endorsingVoters() { 
    return voterReplicaStates
        .values()
        .stream()
        .filter(voter -> voter.hasEndorsedLeader)
        .map(voter -> voter.stateId)
        .collect(Collectors.toSet()); 
}
```

Notice the change in the method name. We can also change `nonEndorsingFollowers` to `nonEndorsingVoters` for consistency.